pkg/apiserver/rest/dualwriter_mode2.go

```diff
@@ -124,28 +126,66 @@ func (d *DualWriterMode2) List(ctx context.Context, options *metainternalversion
 	return ll, nil
 }
 
-func enrichObject(orig, copy runtime.Object) (runtime.Object, error) {
-	accessorC, err := meta.Accessor(copy)
-	if err != nil {
-		return nil, err
-	}
-	accessorO, err := meta.Accessor(orig)
-	if err != nil {
-		return nil, err
-	}
-
+func enrichObject(accessorO, accessorC metav1.Object) {
 	accessorC.SetLabels(accessorO.GetLabels())
 
 	ac := accessorC.GetAnnotations()
 	for k, v := range accessorO.GetAnnotations() {
 		ac[k] = v
 	}
 	accessorC.SetAnnotations(ac)
+}
+
+// Update overrides the behavior of the generic DualWriter and writes first to LegacyStorage and then to Storage.
+func (d *DualWriterMode2) Update(ctx context.Context, name string, objInfo rest.UpdatedObjectInfo, createValidation rest.ValidateObjectFunc, updateValidation rest.ValidateObjectUpdateFunc, forceAllowCreate bool, options *metav1.UpdateOptions) (runtime.Object, bool, error) {
+	legacy, ok := d.Legacy.(rest.Updater)
+	if !ok {
+		return nil, false, errNoUpdateMethod
+	}
 
-	// #TODO set resource version and UID when required (Update for example)
-	// accessorC.SetResourceVersion(accessorO.GetResourceVersion())
+	// get old and new (updated) object so they can be stored in legacy store
+	old, err := d.Get(ctx, name, &metav1.GetOptions{})
+	if err != nil {
+		if !apierrors.IsNotFound(err) {
```

Comment:
I don't think we should be distinguishing between errors here. If we were not able to Get using the DW's Get method that means that the object didn't exist in Storage _and_ in LegacyStorage. Therefore something went wrong and we should return. The current "legacy" behavior is that if there's an issue with an update, we fail without creating an object.

As far as I'm concerned, Dan's [comment](https://github.com/grafana/grafana/pull/85844#discussion_r1561249577) about creating an object refers to the case when an object isn't found in Storage only. The creation of the update in this case should be taken care of by the d.Storage.Update call. Replied in the thread to confirm.

Reply:
I think it's still important to distinguish both errors, but probably it's better just to log different things and still return an error 😄

Reply:
makes sense yeah. Let's do this in a following PR and already with the mock tests setup, to help us spot for edge cases then :)